README

  Welcome to ONS, the "Open Network Suite".

0. Copyright notice
  ONS is developed by the ONS Team.
  This file is part of ONS, see COPYING for details.


                           Table of Contents

                           1. Useful notes
                            2. About ONS
                          3. Installing ONS
                            4. Using ONS
                          5. Developing ONS
                              6. Contact


1. Useful notes
  We are looking for developers which help us improving ONS. If you are
  interested, just contact us (See "6. Contact").

2. About ONS
  The Open Network Suite is an attempt to create a bundle of softwares
  providing well documented access to all network issues. Each software allows
  to use one single network issue and conforms to the Unix ideology to do only
  one thing but do it best. Thus, the softwares provide simple ways to do simple
  things but also complex ways to do complex thing.
  The libraries only sparsely depend on each other, they are provided for
  independant use, however, they are perfect to be used together.

3. Installing ONS
  See INSTALL for install instructions.

4. Using ONS
  See the documentation of ONS and the references of each library to get an
  introduction into ONS. There is also a tutorial available. Though, the
  documentation is the right place for the first steps.

5. Developing ONS
  Please don't hesitate to send patches, hints, ideas or whatever you think
  would enhance ONS to us. We are glad about every submission. Even critics are
  welcome. See "6. Contact" for email addresses.
  If you want to help us improving ONS continuosly, just contact us, too.
  You see, we are glad about every feedback we get.

6. Contact
  Please send a mail to:
    ons@xccd.eu
  or visit us at:
    http://ons.berlios.de
  You can also contact us on an unofficial way by going on IRC and talk
  with one of us in either the query or one of the listed channels. Our nicks
  are listed in COPYING behind the real names.
  (sorted in descending availability)
  - irc.onlinegamesnet.net - #linux or #c++
  - irc.gamesurge.net - #coders
  - irc.quakenet.org - #c or #c.de
  We are neither operators nor well known regulars on most of these channels.
  Best way is just querying us.



ChangeLog
#
#  (COPYRIGHT) Copyright (C) 2008, 2009, The ONS Team.
#  This file is part of ONS, see COPYING for details.
#

#
# File information:
# - Created: 18. May 2008
# - Lead-Dev: - David Herrmann
# - Contributors: /
# - Last-Change: [Last changelog entry]
#

TODO >1.0:
 - asynchio:
    - Add network interface handling (if.h)
    - Add control message handling (cmsg.h)
    - Add all options to options.h.
    - Implement multicast support.
 - misc:
    - config file parser based on trees.

TODO =1.0:
 - build:
    - Write build.bat.
    - Write XY.machine.h configs.
    - Improve autodetection in build.sh.
    - Rewrite INSTALL instructions.
    - Add module selection to build.XY.
    - Add debug selection to build.XY.
 - documentation:
    - complete rules and guidelines
    - Update documentation.
    - ROADMAP
    - add tutorials
 - memoria:
    - make safe for -std=c89 -pedantic -Wshadow
 - asynchio:
    - Add Win/BSD compatibility layer
    - Add ioctl-options support.

TODO on version change:
 - Change AC_INIT in configure.in.
 - Set ONS_DEBUG_MODE in machine.h to a proper value.
 - `grep TODO -R .`
 - Compile on supported platforms with "pedantic".

[ ONS v0.1.1.0 (TESTING) ]
    * David Herrmann 2009/06/26
        * ./.../list.h: Make c89 compatible.
        * ./.../list.c: likewise
        * ./.../splay.c: likewise
        * ./.../rbtree.c: likewise
        * ./.../table.c: Remove the hash table backend. (it sucked)
        * ./.../time.c: Fix global shadow on "ctime".
    * David Herrmann 2009/06/17
        * ./.../machine.h: Change ONS_DEBUG_MODE to have only one level.
    * David Herrmann 2009/06/16
        * ./.../sundry.h: Adjust error messages.
    * David Herrmann 2009/05/27
        * ./code/build.sh: Add "noclean" flag to disable implicit cleaning.
                           Add MODULES variable which can later be used to enable/disable compilation
                           of specific modules.
        * ./code/Makefile: Add asynchio.
                           Force no-pedantic and similar modes on memoria compilation.
        * ./.../sundry.h: Change SUNDRY_FASSERT to SUNDRY_MASSERT because we don't have a "format" argument.
        * ./.../asynchio.h: Implement new asynchio interface.
        *. /.../asynchio/src/*.c: Implement new asynchio interface.
    * David Herrmann 2009/05/26
        * Rewrite build engine. This is hopefully the last time we rewrite it.
    * David Herrmann 2009/05/25
        * ./code/build.sh: Add debug messages and show the user what is actually done.
        * ./README: Update.
        * ./INSTALL: Write install instructions.
        * ./code/config/machine.h: Write configuration instructions.
        * ./code/config/generic.machine.h: create
        * ./code/config/macosx.machine.h: create
    * David Herrmann 2009/04/07
        - Add error codes to ONS.
        - Add new asynchio header/library.
    * David Herrmann 2009/04/04
        - Fix llist to remove elements correctly.
    * David Herrmann 2009/04/03
        - Make hash table dynamic.
        - Fix ONS_CONF_DEBUG to be correctly set in configure.in.
        - Fix mem_list_clear backends again.
        - Make mem_realloc() behave like mem_malloc() if the pointer equals NULL.
    * David Herrmann 2009/04/02
        - Write frontend for a generic double linked list API.
        - Fix bug in mem_list_clear backends.
        - Write hash table list backend.
    * David Herrmann 2009/04/01
        - Write the backend for a generic double linked list API.
    * David Herrmann 2009/03/24
        - Extend mem_array interface to support merges and manual resizing.
        - Rewrite mem_list interface to look always the same and support
          different backends.
    * David Herrmann 2009/03/20
        - Finish new Makefile.in which compiles all libraries into libons.so.
        - Add examples.
        - Fix typos in MEM_DEFINE_ARRAY and extend examples.
        - Fix compoilation on MaxOS-X.
        - Fix example-compilation.
    * David Herrmann 2009/03/15
        - Create new "code" directory and rebuild the whole ONS library with a totally
          new design.
        - Create build scripts for POSIX-SH like environments and WINDOWS-BATCH like
          environments.
    * David Herrmann 2009/02/24
        - Finish main documentation.
        - Finish libons documentation.
        - Add first tutorial.
        - memoria/array.h: Remove the generic array.
                           Change delete function implementation.
        - Fix assertations.
        - memoria/rbtree.c: Fix wrong linked list inserts.
                            Change match function to handle more cases.
        - memoria/splay_tree.c: Implement Splay Trees.
        - saw/net.h: remove
        - fip/error.h: Move error codes into libons.
        - fip: Added debug info printings.
        - yaadns: Update rfc list.
    * David Herrmann 2009/02/23
        - memoria/rbtree.c: Fix key comparison.
                            Add ways to iterate through an RB-tree.
                            Add function to clean a whole rbtree.
                            Link the elements additionally in a double linked list which is
                            not circular, but straight-lined.
        - memoria/splay_tree.[hc]: Add splay-tree primitives.
        - memoria/llist.h: Change link list to a straight-lined linked list.
    * David Herrmann 2009/02/23
        - memoria/array.h: Fix MEM_ARRAY_DEFINE typo.
        - memoria/rbtree.[hc]: Implement Red-Black Trees.
    * David Herrmann 2009/02/22
        - memoria/llist.h: Change \count member to "size_t". This guarantees that it
                           cannot overflow.
        - memoria/misc.h: Change "mem_index_t" to "size_t".
        - libmemoria.doc.xhtml: Finish llist documentation.
        - COPYING: License ONS under a second license: ISC license.
        - memoria/hash.h: Fix hash table comments.
        - memoria/rbtree.[hc]: Add Red-Black Tree files.

code/memoria/src/hash.c
/*
 * (COPYRIGHT) Copyright (C) 2008, 2009, The ONS Team.
 * This file is part of ONS, see COPYING for details.
 */

/*
 * File information:
 * - Created: 20. December 2008
 * - Lead-Dev: - David Herrmann
 * - Contributors: /
 * - Last-Change: 26. May 2009
 */

/* Hash list implementation.
 * The hash algorithm is based on the one created by Bob Jenkins. It is dedicated
 * to the public domain. We thank him for the effort he took to create this
 * nice piece of software:
 * -------------------------------------------------------------------------------
 * lookup3.c, by Bob Jenkins, May 2006, Public Domain.
 * -------------------------------------------------------------------------------
 * This file is a modified lookup3.c.
 */


#include "config/machine.h"
#include "memoria/memoria.h"
#include "memoria/alloc.h"
#include "memoria/array.h"


/** Hash Algorithm.
 * This part implements a very fast hash algorithm developed by Bob Jenkins under
 * the public domain. It is divided into a big and little endian part. Only one of
 * them is available on your machine.
 * The comments in this part are mostly copied identically from Bob Jenkins, however,
 * the functions are slightly modified to be ISO-C compatible and to be formatted
 * and indented in a clear manner.
 */


/* Mixes three 32bit values reversibly.
 * This is reversible, so any information in (a, b, c) before mem_mix() is
 * still in (a, b, c) after mem_mix().
 *
 * Macro originally created by Bob Jenkins (public domain).
 *
 * If four pairs of (a, b, c) inputs are run through mem_mix(), or through
 * mem_mix() in reverse, there are at least 32 bits of the output that
 * are sometimes the same for one pair and different for another pair.
 * This was tested for:
 * * pairs that differed by one bit, by two bits, in any combination
 *   of top bits of (a, b, c), or in any combination of bottom bits of
 *   (a, b, c).
 * * "differ" is defined as +, -, ^, or ~^.  For + and -, I transformed
 *   the output delta to a Gray code (a^(a>>1)) so a string of 1's (as
 *   is commonly produced by subtraction) look like a single 1-bit
 *   difference.
 * * the base values were pseudorandom, all zero but one bit set, or 
 *   all zero plus a counter that starts at zero.
 *
 * Some k values for my "a-=c; a^=rot(c,k); c+=b;" arrangement that
 * satisfy this are
 *     4  6  8 16 19  4
 *     9 15  3 18 27 15
 *    14  9  3  7 17  3
 * Well, "9 15 3 18 27 15" didn't quite get 32 bits diffing
 * for "differ" defined as + with a one-bit base and a two-bit delta. I
 * used http://burtleburtle.net/bob/hash/avalanche.html to choose 
 * the operations, constants, and arrangements of the variables.
 *
 * This does not achieve avalanche. There are input bits of (a, b, c)
 * that fail to affect some output bits of (a, b, c), especially of a. The
 * most thoroughly mixed value is c, but it doesn't really even achieve
 * avalanche in c.
 *
 * This allows some parallelism. Read-after-writes are good at doubling
 * the number of bits affected, so the goal of mixing pulls in the opposite
 * direction as the goal of parallelism. I did what I could. Rotates
 * seem to cost as much as shifts on every machine I could lay my hands
 * on, and rotates are much kinder to the top and bottom bits, so I used
 * rotates.
 */
#define mem_mix(a, b, c) \
{ \
    a -= c;  a ^= MEM_LROT(32, c, 4);  c += b; \
    b -= a;  b ^= MEM_LROT(32, a, 6);  a += c; \
    c -= b;  c ^= MEM_LROT(32, b, 8);  b += a; \
    a -= c;  a ^= MEM_LROT(32, c,16);  c += b; \
    b -= a;  b ^= MEM_LROT(32, a,19);  a += c; \
    c -= b;  c ^= MEM_LROT(32, b, 4);  b += a; \
}

/* final -- final mixing of 3 32-bit values (a,b,c) into c
 * Macro originally created by Bob Jenkins (public domain).
 *
 * Pairs of (a,b,c) values differing in only a few bits will usually
 * produce values of c that look totally different.  This was tested for
 * * pairs that differed by one bit, by two bits, in any combination
 *   of top bits of (a,b,c), or in any combination of bottom bits of
 *   (a,b,c).
 * * "differ" is defined as +, -, ^, or ~^.  For + and -, I transformed
 *   the output delta to a Gray code (a^(a>>1)) so a string of 1's (as
 *   is commonly produced by subtraction) look like a single 1-bit
 *   difference.
 * * the base values were pseudorandom, all zero but one bit set, or 
 *   all zero plus a counter that starts at zero.
 *
 * These constants passed:
 *  14 11 25 16 4 14 24
 *  12 14 25 16 4 14 24
 * and these came close:
 *   4  8 15 26 3 22 24
 *  10  8 15 26 3 22 24
 *  11  8 15 26 3 22 24
 */
#define mem_final(a, b, c) \
{ \
    c ^= b; c -= MEM_LROT(32, b,14); \
    a ^= c; a -= MEM_LROT(32, c,11); \
    b ^= a; b -= MEM_LROT(32, a,25); \
    c ^= b; c -= MEM_LROT(32, b,16); \
    a ^= c; a -= MEM_LROT(32, c,4);  \
    b ^= a; b -= MEM_LROT(32, a,14); \
    c ^= b; c -= MEM_LROT(32, b,24); \
}

/** mem_hashlittle()
 * This function should only be available on little endian machines.
 * mem_hashbig() does the same on big endian machines. Both produce
 * the same output.
 * Do not make the function visible to the user.
 */
/* mem_hashlittle() -- hash a variable-length key into a 32-bit value
 * Function originally created by Bob Jenkins (public domain).
 *
 *  k       : the key (the unaligned variable-length array of bytes)
 *  length  : the length of the key, counting by bytes
 *  initval : can be any 4-byte value
 * Returns a 32-bit value.  Every bit of the key affects every bit of
 * the return value.  Two keys differing by one or two bits will have
 * totally different hash values.
 *
 * The best hash table sizes are powers of 2.  There is no need to do
 * mod a prime (mod is sooo slow!).  If you need less than 32 bits,
 * use a bitmask.  For example, if you need only 10 bits, do
 *   h = (h & hashmask(10));
 * In which case, the hash table should have hashsize(10) elements.
 *
 * If you are hashing n strings (uint8_t **)k, do it like this:
 *   for (i=0, h=0; i<n; ++i) h = hashlittle( k[i], len[i], h);
 *
 * Use for hash table lookup, or anything where one collision in 2^^32 is
 * acceptable.  Do NOT use for cryptographic purposes.
 */
#ifdef ONS_ARCH_LITTLEENDIAN
static uint32_t mem_hashlittle(const void *key, size_t length, uint32_t initval) {
    uint32_t a, b, c; /* internal state */
    union { const void *ptr; size_t i; } u; /* needed for Mac Powerbook G4 */

    SUNDRY_ASSERT(key != NULL);
    SUNDRY_ASSERT(length > 0);

    /* Set up the internal state */
    a = b = c = 0xdeadbeef + ((uint32_t)length) + initval;

    u.ptr = key;
    if((u.i & 0x3) == 0) {
        const uint32_t *k = (const uint32_t *)key; /* read 32-bit chunks */
        const uint8_t *k8;

        /* All but last block: Aligned reads and affect 32 bits of (a,b,c). */
        while (length > 12) {
            a += k[0];
            b += k[1];
            c += k[2];
            mem_mix(a, b, c);
            length -= 12;
            k += 3;
        }

        k8 = (const uint8_t *)k;
        switch(length) {
            case 12:
                c += k[2];
                b += k[1];
                a += k[0];
                break;
            case 11:
                c += ((uint32_t)k8[10]) << 16;
            case 10:
                c += ((uint32_t)k8[9]) << 8;
            case 9:
                c += k8[8];
            case 8:
                b += k[1];
                a += k[0];
                break;
            case 7:
                b += ((uint32_t)k8[6]) << 16;
            case 6:
                b += ((uint32_t)k8[5]) << 8;
            case 5:
                b += k8[4];
            case 4:
                a += k[0];
                break;
            case 3:
                a += ((uint32_t)k8[2]) << 16;
            case 2:
                a += ((uint32_t)k8[1]) << 8;
            case 1:
                a += k8[0];
                break;
            case 0:
                return c;
        }
    }
    else if((u.i & 0x1) == 0) {
        const uint16_t *k = (const uint16_t *)key; /* read 16-bit chunks */
        const uint8_t  *k8;

        /* All but last block: Aligned reads and different mixing. */
        while(length > 12) {
            a += k[0] + (((uint32_t)k[1]) << 16);
            b += k[2] + (((uint32_t)k[3]) << 16);
            c += k[4] + (((uint32_t)k[5]) << 16);
            mem_mix(a, b, c);
            length -= 12;
            k += 6;
        }

        /* Handle the last (probably partial) block. */
        k8 = (const uint8_t *)k;
        switch(length) {
            case 12: c += k[4] + (((uint32_t)k[5]) << 16);
                     b += k[2] + (((uint32_t)k[3]) << 16);
                     a += k[0] + (((uint32_t)k[1]) << 16);
                     break;
            case 11: c += ((uint32_t)k8[10]) << 16;
            case 10: c += k[4];
                     b += k[2] + (((uint32_t)k[3]) << 16);
                     a += k[0] + (((uint32_t)k[1]) << 16);
                     break;
            case 9 : c += k8[8];
            case 8 : b += k[2] + (((uint32_t)k[3]) << 16);
                     a += k[0] + (((uint32_t)k[1]) << 16);
                     break;
            case 7 : b += ((uint32_t)k8[6]) << 16;
            case 6 : b += k[2];
                     a += k[0] + (((uint32_t)k[1]) << 16);
                     break;
            case 5 : b += k8[4];
            case 4 : a += k[0] + (((uint32_t)k[1]) << 16);
                     break;
            case 3 : a += ((uint32_t)k8[2]) << 16;
            case 2 : a += k[0];
                     break;
            case 1 : a += k8[0];
                     break;
            case 0 : return c;                     /* zero length requires no mixing */
        }
    }
    else { /* need to read the key one byte at a time */
        const uint8_t *k = (const uint8_t *)key;

        /* All but the last block: Affect some 32 bits of (a,b,c). */
        while(length > 12) {
            a += k[0];
            a += ((uint32_t)k[1]) << 8;
            a += ((uint32_t)k[2]) << 16;
            a += ((uint32_t)k[3]) << 24;
            b += k[4];
            b += ((uint32_t)k[5]) << 8;
            b += ((uint32_t)k[6]) << 16;
            b += ((uint32_t)k[7]) << 24;
            c += k[8];
            c += ((uint32_t)k[9]) << 8;
            c += ((uint32_t)k[10]) << 16;
            c += ((uint32_t)k[11]) << 24;
            mem_mix(a, b, c);
            length -= 12;
            k += 12;
        }

        /* Last block: Affect all 32 bits of (c). */
        switch(length) {
            case 12: c += ((uint32_t)k[11]) << 24;
            case 11: c += ((uint32_t)k[10]) << 16;
            case 10: c += ((uint32_t)k[9]) << 8;
            case 9 : c += k[8];
            case 8 : b += ((uint32_t)k[7]) << 24;
            case 7 : b += ((uint32_t)k[6]) << 16;
            case 6 : b += ((uint32_t)k[5]) << 8;
            case 5 : b += k[4];
            case 4 : a += ((uint32_t)k[3]) << 24;
            case 3 : a += ((uint32_t)k[2]) << 16;
            case 2 : a += ((uint32_t)k[1]) << 8;
            case 1 : a += k[0];
                     break;
            case 0 : return c;
        }
  }

  mem_final(a, b, c);
  return c;
}
#endif /* ONS_ARCH_LITTLEENDIAN */

/** mem_hashbig()
 * This function should only be available on big endian machines.
 * mem_hashlittle() does the same on little endian machines. Both produce
 * the same output.
 * Do not make the function visible to the user.
 */
/*
 * hashbig():
 * This is the same as hashlittle() but on big-endian machines.
 * hashbig() takes advantage of big-endian byte ordering.
 */
#ifdef ONS_ARCH_BIGENDIAN
static uint32_t mem_hashbig(const void *key, size_t length, uint32_t initval) {
    uint32_t a, b, c;
    union { const void *ptr; size_t i; } u; /* to cast key to (size_t) happily */

    SUNDRY_ASSERT(key != NULL);
    SUNDRY_ASSERT(length > 0);

    /* Set up the internal state */
    a = b = c = 0xdeadbeef + ((uint32_t)length) + initval;

    u.ptr = key;
    if((u.i & 0x3) == 0) {
        const uint32_t *k = (const uint32_t *)key; /* read 32-bit chunks */
        const uint8_t  *k8;

        while(length > 12) {
            a += k[0];
            b += k[1];
            c += k[2];
            mem_mix(a, b, c);
            length -= 12;
            k += 3;
        }

        k8 = (const uint8_t *)k;
        switch(length) {
            case 12: c += k[2];
                     b += k[1];
                     a += k[0];
                     break;
            case 11: c += ((uint32_t)k8[10]) << 8;
            case 10: c += ((uint32_t)k8[9]) << 16;
            case 9 : c += ((uint32_t)k8[8]) << 24;
            case 8 : b += k[1];
                     a += k[0];
                     break;
            case 7 : b += ((uint32_t)k8[6]) << 8;
            case 6 : b += ((uint32_t)k8[5]) << 16;
            case 5 : b += ((uint32_t)k8[4]) << 24;
            case 4 : a += k[0];
                     break;
            case 3 : a += ((uint32_t)k8[2]) << 8;
            case 2 : a += ((uint32_t)k8[1]) << 16;
            case 1 : a += ((uint32_t)k8[0]) << 24;
                     break;
            case 0 : return c;
        }
    }
    else { /* need to read the key one byte at a time */
        const uint8_t *k = (const uint8_t *)key;

        while(length > 12) {
            a += ((uint32_t)k[0]) << 24;
            a += ((uint32_t)k[1]) << 16;
            a += ((uint32_t)k[2]) << 8;
            a += ((uint32_t)k[3]);
            b += ((uint32_t)k[4]) << 24;
            b += ((uint32_t)k[5]) << 16;
            b += ((uint32_t)k[6]) << 8;
            b += ((uint32_t)k[7]);
            c += ((uint32_t)k[8]) << 24;
            c += ((uint32_t)k[9]) << 16;
            c += ((uint32_t)k[10]) << 8;
            c += ((uint32_t)k[11]);
            mem_mix(a, b, c);
            length -= 12;
            k += 12;
        }

        switch(length) {
            case 12: c += k[11];
            case 11: c += ((uint32_t)k[10]) << 8;
            case 10: c += ((uint32_t)k[9]) << 16;
            case 9 : c += ((uint32_t)k[8]) << 24;
            case 8 : b += k[7];
            case 7 : b += ((uint32_t)k[6]) << 8;
            case 6 : b += ((uint32_t)k[5]) << 16;
            case 5 : b += ((uint32_t)k[4]) << 24;
            case 4 : a += k[3];
            case 3 : a += ((uint32_t)k[2]) << 8;
            case 2 : a += ((uint32_t)k[1]) << 16;
            case 1 : a += ((uint32_t)k[0]) << 24;
                     break;
            case 0 : return c;
        }
    }
    mem_final(a, b, c);
    return c;
}
#endif /* ONS_ARCH_BIGENDIAN */

/* Wrapper for big and little endian hash functions.
 * Takes as argument a char array and a length argument
 * and returns a 32bit hash value. This value is not safe
 * for cryptografic issues.
 * It is a very fast algorithm, though, it has really little
 * collisions and is almost perfectly balanced.
 */
uint32_t mem_hash(const char *str, size_t len) {
    SUNDRY_ASSERT(str != NULL);
    SUNDRY_ASSERT(len > 0);

#ifdef ONS_ARCH_LITTLEENDIAN
    return mem_hashlittle(str, len, 0xdefcadad);
#elif defined(ONS_ARCH_BIGENDIAN)
    return mem_hashbig(str, len, 0xdefcadad);
#else
    #error "Your machine must either be little- or bigendian. Please review your configuration."
    return 0;
#endif
}

